sound/soc/sof/card-tplg-ops.c

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2025 Intel Corporation.
+//
+
+#include <linux/device.h>
+#include <linux/errno.h>
+#include "sof-priv.h"
+#include "topology.h"
+
+enum tplg_device_id {
+	TPLG_DEVICE_SDW_JACK,
+	TPLG_DEVICE_SDW_AMP,
+	TPLG_DEVICE_SDW_MIC,
+	TPLG_DEVICE_PCH_DMIC,
+	TPLG_DEVICE_HDMI,
+	TPLG_DEVICE_MAX
+};
+
+#define SDCA_DEVICE_MASK (BIT(TPLG_DEVICE_SDW_JACK) | BIT(TPLG_DEVICE_SDW_AMP) | \
+			  BIT(TPLG_DEVICE_SDW_MIC))
+
+#define PLATFORM_NAME_MAX 4
+
+int get_sof_sdw_tplg_files(struct snd_soc_component *scomp, const char *file, char ***tplg_files)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
+	struct snd_soc_dai_link *dai_link;
+	char platform[PLATFORM_NAME_MAX];
+	unsigned long tplg_mask = 0;
+	int tplg_num = 0;
+	int tplg_dev;
+	int ret;
+	int i;
+
+	if (sdev->pdata->ipc_type == SOF_IPC_TYPE_3)
+		return 0;
+
+	ret = sscanf(sof_pdata->tplg_filename, "sof-%3s-*.tplg", platform);
+	if (ret != 1)
+		return 0;
+
+	/* tplg_num is always less or equal to dai_link numbers */
+	*tplg_files = kcalloc(scomp->card->num_links, sizeof(char *), GFP_KERNEL);
+	if (!(*tplg_files))
+		return -ENOMEM;
+
+	for_each_card_prelinks(scomp->card, i, dai_link) {
+		char *tplg_device;
+
+		dev_dbg(scomp->dev, "dai_link %s id %d\n", dai_link->name, dai_link->id);
+		if (strstr(dai_link->name, "SimpleJack")) {
+			tplg_dev = TPLG_DEVICE_SDW_JACK;
+			tplg_device = "sdca-jack";
+		} else if (strstr(dai_link->name, "SmartAmp")) {
+			tplg_dev = TPLG_DEVICE_SDW_AMP;
+			tplg_device = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						     "sdca-%damp", dai_link->num_cpus);
+			if (!tplg_device)
+				return -ENOMEM;
+		} else if (strstr(dai_link->name, "SmartMic")) {
+			tplg_dev = TPLG_DEVICE_SDW_MIC;
+			tplg_device = "sdca-mic";
+		} else if (strstr(dai_link->name, "dmic")) {
+			if (strstr(file, "-2ch")) {
+				tplg_device = "dmic-2ch";
+			} else if (strstr(file, "-4ch")) {
+				tplg_device = "dmic-4ch";
+			} else {
+				dev_warn(scomp->dev,
+					 "only -2ch and -4ch are supported for dmic\n");
+				continue;
+			}
+			tplg_dev = TPLG_DEVICE_PCH_DMIC;
+		} else if (strstr(dai_link->name, "iDisp")) {
+			tplg_dev = TPLG_DEVICE_HDMI;
+			tplg_device = "hdmi-pcm5";
+
+		} else {
+			/* The dai link is not supported by sperated tplg yet */
+			dev_dbg(scomp->dev,
+				"dai_link %s is not supported by sperated tplg yet, Fail back to %s\n",
+				dai_link->name, file);
+			return 0;
+		}
+		if (tplg_mask & BIT(tplg_dev))
+			continue;
+
+		tplg_mask |= BIT(tplg_dev);
+		switch (tplg_dev) {
+		case TPLG_DEVICE_PCH_DMIC:
+			(*tplg_files)[tplg_num] = kasprintf(GFP_KERNEL, "%s/sof-%s-%s-id%d.tplg",
```

Comment:
you've allocated `*tplg_files` as just a single pointer, and here you're using it as a virtually endless (depending on the number of the cards) array. This seems like a memory corruption to me.

Reply:
@bardliao oh, you're right, my oversight, sorry